pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/LookupTransformFunction.java

```diff
@@ -0,0 +1,284 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.core.operator.transform.function;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.ArrayList;
+import com.google.common.base.Preconditions;
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.pinot.core.common.DataSource;
+import org.apache.pinot.core.data.manager.offline.DimensionTableDataManager;
+import org.apache.pinot.core.operator.blocks.ProjectionBlock;
+import org.apache.pinot.core.operator.transform.TransformResultMetadata;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.apache.pinot.spi.data.readers.PrimaryKey;
+
+/**
+ * LOOKUP function take 4 or more arguments:
+ * <ul>
+ *   <li><b>TableName:</b> name of the dimension table which will be used</li>
+ *   <li><b>ColumnName:</b> column name from the dimension table to look up</li>
+ *   <li><b>JoinKey:</b> primary key column name for the dimension table. Note: Only primary key[s] is supported for JoinKey</li>
+ *   <li><b>JoinValue:</b> primary key value</li>
+ *   ...<br>
+ *   *[If the dimension table has more then one primary keys (composite pk)]
+ *     <li><b>JoinKey2</b></li>
+ *     <li><b>JoinValue2</b></li>
+ *   ...
+ * </ul>
+ * <br>
+ * Example:
+ * <pre>{@code select playerName, teamID, lookup('baseballTeams', 'teamName', 'teamID', teamID) from baseballStats limit 10}</pre>
+ * <br>
+ * Above example joins the dimension table 'baseballTeams' into regular table 'baseballStats' on 'teamID' key.
+ * Lookup function returns the value of the column 'teamName'.
+ */
+public class LookupTransformFunction extends BaseTransformFunction {
+    public static final String FUNCTION_NAME = "lookUp";
+    private static final String TABLE_NAME_SUFFIX = "_OFFLINE";
```

Comment:
I feel we don't need to assume a dim table is always an offline table?

Reply:
Hi Xiang, yes you're right. In current design Dimension Table has the following constraints:
- Must be of OFFLINE type
- Must have a primary key (we support lookups by primary key for now)
- Must have ingestion type REFRESH

Please check out #6286 and #6346 to see implementation details.